### PR TITLE
Fix recipe selection

### LINKED ISSUE #2172: "latex-workshop.latex.recipe.default": "lastUsed" doesn't work
LaTeX-Workshop not honoring "latex-workshop.latex.recipe.default": "lastUsed" when building.

## Describe the bug [Required]

LaTeX-Workshop not honoring "latex-workshop.latex.recipe.default": "lastUsed" when building.
### To Reproduce

Build a project using 'xelatex' engine recipe.
Make modifications to the project.
Save the file.
Run Build LaTeX project (or save the file if latex-workshop.latex.autoBuild.run equals onFileChange).

### Expected behavior

The 'xelatex' recipe should be run but the 'pdfLatex' recipe run.

## Logs
```
[13:21:26] Found root file from active editor: /home/rio/Documents/HW4_AI_9512058/HW4_AI_9512058.tex
[13:21:26] Keep using the same root file: /home/rio/Documents/HW4_AI_9512058/HW4_AI_9512058.tex.
[13:21:31] Formatting with command /usr/local/texlive/2020/texmf-dist/scripts/latexindent/latexindent.pl -c,/home/rio/Documents/HW4_AI_9512058/,/home/rio/Documents/HW4_AI_9512058/__latexindent_temp.tex,-y=defaultIndent: '    '
[13:21:32] Formatted /home/rio/Documents/HW4_AI_9512058/HW4_AI_9512058.tex
[13:21:32] File watcher - file changed: /home/rio/Documents/HW4_AI_9512058/HW4_AI_9512058.tex
[13:21:32] Parsing a file and its subfiles: /home/rio/Documents/HW4_AI_9512058/HW4_AI_9512058.tex
[13:21:32] Parse fls file.
[13:21:32] Cannot find fls file: /home/rio/Documents/HW4_AI_9512058/HW4_AI_9512058.fls
[13:21:32] Auto build started detecting the change of a file: /home/rio/Documents/HW4_AI_9512058/HW4_AI_9512058.tex
[13:21:32] BUILD command invoked.
[13:21:32] Building root file: /home/rio/Documents/HW4_AI_9512058/HW4_AI_9512058.tex
[13:21:32] Build root file /home/rio/Documents/HW4_AI_9512058/HW4_AI_9512058.tex
[13:21:32] Recipe step 1: pdflatex, -synctex=1,-interaction=nonstopmode,-file-line-error,/home/rio/Documents/HW4_AI_9512058/HW4_AI_9512058
[13:21:32] cwd: /home/rio/Documents/HW4_AI_9512058
[13:21:32] LaTeX build process spawned. PID: 12271.
[13:21:32] Recipe returns with error: 1/null. PID: 12271. message: .
[13:21:32] Cleaning auxiliary files and retrying build after toolchain error.
[13:21:32] File cleaned: /home/rio/Documents/HW4_AI_9512058/HW4_AI_9512058.log
[13:21:32] Recipe step 1: pdflatex, -synctex=1,-interaction=nonstopmode,-file-line-error,/home/rio/Documents/HW4_AI_9512058/HW4_AI_9512058
[13:21:32] cwd: /home/rio/Documents/HW4_AI_9512058
[13:21:32] LaTeX build process spawned. PID: 12272.
[13:21:33] Recipe returns with error: 1/null. PID: 12272. message: .
```

 - OS: Ubuntu 18.04
 - VS Code version: 1.46.1
 - Extension version: 8.11.0
 - TeX distribution version: TeX Live 2020

### PR DESCRIPTION
**Pull request by jlelong**
Fix #2172

### CODE REVIEW

File: src/components/builder.ts [hunk 1]
```diff
@@ -403,21 +408,19 @@ export class Builder {
             }
         } else {
             const recipes = configuration.get('latex.recipes') as {name: string, tools: (string | StepCommand)[]}[]
+            const defaultRecipe = configuration.get('latex.recipe.default') as string
```
Comment (tamuratak) [inline src/components/builder.ts:411]: `defaultRecipeName` might be better for the variable name.

File: src/components/builder.ts [hunk 2]
```diff
@@ -426,18 +429,27 @@ export class Builder {
...
+                if (defaultRecipe === 'first' || recipe === undefined) {
+                   let candidates: Recipe[] = recipes
+                   if (languageId === 'rsweave') {
+                        candidates = recipes.filter(candidate => candidate.name.toLowerCase().match('rnw|rsweave'))
+                   } else if (languageId === 'jlweave') {
+                        candidates = recipes.filter(candidate => candidate.name.toLowerCase().match('jnw|jlweave|weave.jl'))
+                   }
+                    if (candidates.length < 1) {
+                        this.extension.logger.addLogMessage(`Failed to resolve build recipe: ${recipeName}`)
+                        this.extension.logger.showErrorMessage(`Failed to resolve build recipe: ${recipeName}`)
```
Comment (tamuratak) [inline src/components/builder.ts:446]: We should immediately return.

File: src/components/builder.ts [hunk 3]
```diff
@@ -426,18 +429,27 @@ export class Builder {
                     this.extension.logger.showErrorMessage(`Failed to resolve build recipe: ${recipeName}`)
```
Comment (tamuratak) [inline src/components/builder.ts:429]: We should immediately return.

**Comment by jlelong** [comment]
@tamuratak at some point you suggested to immediately return if the default recipe name is not found, but I cannot find the posts anymore. Have you changed your mind?

**Comment by tamuratak** [comment]
Yes, I have removed them. They are not necessary.

File: src/components/builder.ts [hunk 4]
```diff
@@ -403,21 +408,19 @@ export class Builder {
             }
         } else {
             const recipes = configuration.get('latex.recipes') as {name: string, tools: (string | StepCommand)[]}[]
```
Comment (tamuratak) [inline src/components/builder.ts:410]: You can use the interface `Recipe` you have defined, here.

**Comment by tamuratak** [comment]
I do not see any issue.

**Pull request merged**